iceoryx_dds/source/iceoryx_dds/dds/cyclone_data_reader.cpp

```diff
@@ -13,7 +13,190 @@
 // limitations under the License.
 
 #include "iceoryx_dds/dds/cyclone_data_reader.hpp"
+#include "iceoryx_dds/dds/cyclone_context.hpp"
+#include "iceoryx_dds/internal/log/logging.hpp"
 
 iox::dds::CycloneDataReader::CycloneDataReader(IdString serviceId, IdString instanceId, IdString eventId) noexcept
+    : m_serviceId(serviceId)
+    , m_instanceId(instanceId)
+    , m_eventId(eventId)
 {
+    LogDebug() << "[CycloneDataReader] Created CycloneDataReader.";
+}
+
+iox::dds::CycloneDataReader::~CycloneDataReader()
+{
+    LogDebug() << "[CycloneDataReader] Destroyed CycloneDataReader.";
+}
+
+void iox::dds::CycloneDataReader::connect() noexcept
+{
+    if (!m_isConnected.load(std::memory_order_relaxed))
+    {
+        auto topicString =
+            "/" + std::string(m_serviceId) + "/" + std::string(m_instanceId) + "/" + std::string(m_eventId);
+        auto topic = ::dds::topic::Topic<Mempool::Chunk>(CycloneContext::getParticipant(), topicString);
+        auto subscriber = ::dds::sub::Subscriber(CycloneContext::getParticipant());
+
+        auto qos = ::dds::sub::qos::DataReaderQos();
+        qos << ::dds::core::policy::History::KeepAll();
+
+        m_impl = ::dds::sub::DataReader<Mempool::Chunk>(subscriber, topic, qos);
+
+        LogDebug() << "[CycloneDataReader] Connected to topic: " << topicString;
+
+        m_isConnected.store(true, std::memory_order_relaxed);
+    }
+}
+
+iox::cxx::optional<uint64_t> iox::dds::CycloneDataReader::peekNextSize()
+{
+    // ensure to only read sample - do not take
+    auto readSamples = m_impl.select()
+                       .max_samples(1u)
+                       .state(::dds::sub::status::SampleState::any())
+                       .read();
+
+    if(readSamples.length() > 0)
+    {
+        auto nextSample = readSamples.begin();
+        auto nextSampleSize = nextSample->data().payload().size();
+
+        // Ignore samples with no payload
+        if (nextSampleSize != 0)
+        {
+            return iox::cxx::optional<uint64_t>(static_cast<uint64_t>(nextSampleSize));
+        }
+    }
+
+    // no valid samples available
+    return iox::cxx::nullopt_t();
+
+}
+
+iox::cxx::expected<iox::dds::DataReaderError>
+iox::dds::CycloneDataReader::takeNext(uint8_t* const buffer, const uint64_t& bufferSize)
```

Comment:
Have you considered using a `std::span` at these kind of signatures? https://en.cppreference.com/w/cpp/container/span It's a lot safer than a pointer+size API and avoids out of bounds accesses. It could be easily implemented as part of iceoryx_utils.

Did you have any other places with this kind of API in your DDS story?

Reply:
Create #180 for this point.